external-crates/move/crates/move-compiler/src/naming/syntax_methods.rs

```diff
@@ -0,0 +1,514 @@
+// Resolution and recording
+//-------------------------------------------------------------------------------------------------
+
+/// validate and record syntax methods
+pub(in crate::naming) fn resolve_syntax_attributes(
+    context: &mut Context,
+    syntax_methods: &mut SyntaxMethods,
+    module_name: &ModuleIdent,
+    function_name: &FunctionName,
+    function: &N::Function,
+) -> Option<()> {
+    let attr = function.attributes.get_(&SyntaxAttribute::Syntax.into())?;
+    let attr_loc = attr.loc;
+
+    let syntax_method_prekinds = resolve_syntax_method_prekind(context.env, attr)?;
+
+    if !context.env.check_feature(
+        context.current_package,
+        FeatureGate::SyntaxMethods,
+        attr_loc,
+    ) {
+        return None;
+    }
+
+    let param_ty = get_first_type(context, &attr_loc, &function.signature)?;
+    let Some(type_name) = determine_subject_type_name(context, module_name, &attr_loc, &param_ty)
+    else {
+        assert!(context.env.has_errors());
+        return None;
+    };
+
+    if syntax_method_prekinds.is_empty() {
+        assert!(context.env.has_errors());
+        return None;
+    }
+
+    // For loops may need to change this, but for now we disallow this.
+    if let Some(macro_loc) = function.macro_ {
+        let msg = "Syntax attributes may not appear on macro definitions";
+        let fn_msg = "This function is a macro";
+        context.env.add_diag(diag!(
+            Declarations::InvalidSyntaxMethod,
+            (attr_loc, msg),
+            (macro_loc, fn_msg)
+        ));
+        return None;
+    }
+
+    let public_visibility = match function.visibility {
+        E::Visibility::Public(loc) => loc,
+        E::Visibility::Friend(_) | E::Visibility::Package(_) | E::Visibility::Internal => {
+            let msg = "Syntax attributes may only appear on public functions";
+            let fn_msg = "This function is not public";
+            context.env.add_diag(diag!(
+                Declarations::InvalidVisibilityModifier,
+                (attr_loc, msg),
+                (function_name.0.loc, fn_msg)
+            ));
+            return None;
+        }
+    };
+
+    let method_entry = syntax_methods.entry(type_name.clone()).or_default();
+
+    for prekind in syntax_method_prekinds {
+        let Some(kind) = determine_valid_kind(context, prekind, &param_ty) else {
+            assert!(context.env.has_errors());
+            continue;
+        };
+        if !valid_return_type(context, &kind, &function.signature.return_type) {
+            assert!(context.env.has_errors());
+            continue;
+        }
+        let new_syntax_method = SyntaxMethod {
+            loc: function_name.0.loc,
+            public_visibility,
+            kind,
+            tname: type_name.clone(),
+            target_function: (*module_name, *function_name),
+        };
+        let kind_loc: &mut Option<Box<SyntaxMethod>> = method_entry.lookup_kind_entry(&kind);
```

Comment:
Maybe 
```suggestion
        let method_opt: &mut Option<Box<SyntaxMethod>> = method_entry.lookup_kind_entry(&kind);
```
Or something?